files/galaxy/dynamic_job_rules/production/total_perspective_vortex/default_tool.yml.j2

```diff
@@ -65,15 +66,33 @@ tools:
 
           def destination_usage_proportion(destination):
             if not destination.context.get('destination_total_mem') or not destination.context.get('destination_total_cores'):
-                    raise Exception(f"++ ---tpv rank debug: At least one of destination_total_mem, destination_total_cores is unavailable")
+              raise Exception(f"++ ---tpv rank debug: At least one of destination_total_mem, destination_total_cores is unavailable")
             destination_total_cores = destination.context.get('destination_total_cores')
             return_value = sum([db_queue_info.get(destination.id, {}).get(state, {}).get('sum_cores', 0) for state in ['queued', 'running']])/destination_total_cores
```

Comment:
Since state 'queued' and 'running' are already filtered out at the sql level, and both states are always combined into a single value, is there an advantage in breaking things down by queued and running in the db_queue_info dict?

Reply:
No, that is a good point

Reply:
part of the reason for this is that I would like to write a generic helper function for tpv around the parameters 'tpv_cores', 'tpv_mem', 'tpv_gpus'. I am not considering a difference between queued and running but someone else might, and the return value would include that information

```
def get_queue_info(app):
    from sqlalchemy import func, text, Float
    from galaxy.model import Job
    query = app.model.context.query(
        Job.destination_id,
        Job.state,
        func.count(Job.id).label('job_count'),
        func.sum(
            func.cast(text("encode(destination_params, 'escape')::json ->> 'tpv_cores'"), Float)
        ).label('total_cores'),
        func.sum(
            func.cast(text("encode(destination_params, 'escape')::json ->> 'tpv_mem'"), Float)
        ).label('total_mem'),
        func.sum(
            func.cast(text("encode(destination_params, 'escape')::json ->> 'tpv_gpus'"), Float)
        ).label('total_gpus'),
    ).filter(
        Job.state.in_(['queued', 'running'])
    ).group_by(
        Job.destination_id,
        Job.state
    )
    results = query.all()
    < construct db_queue_info from results and return db_queue_info >
```

Reply:
Right, that makes sense. Let's leave it as is then. We could also consider making the job states a parameter to get_queue_info.